File: backend/src/utils/auth_utils.py
import os
from datetime import datetime, timedelta, timezone

import bcrypt
import jwt

JWT_SECRET = os.getenv("JWT_SECRET", "change-me-in-production")
JWT_ALGORITHM = "HS256"
JWT_EXPIRATION_HOURS = 24

# bcrypt has a 72-byte limit; truncate to avoid ValueError
BCRYPT_MAX_BYTES = 72


def _to_bcrypt_input(password: str) -> bytes:
    encoded = password.encode("utf-8")
    return encoded[:BCRYPT_MAX_BYTES] if len(encoded) > BCRYPT_MAX_BYTES else encoded


def hash_password(password: str) -> str:
    pwd_bytes = _to_bcrypt_input(password)
    return bcrypt.hashpw(pwd_bytes, bcrypt.gensalt()).decode("utf-8")


def verify_password(plain_password: str, hashed_password: str) -> bool:
    pwd_bytes = _to_bcrypt_input(plain_password)
    return bcrypt.checkpw(pwd_bytes, hashed_password.encode("utf-8"))


def create_access_token(data: dict) -> str:
    payload = data.copy()
    expire = datetime.now(timezone.utc) + timedelta(hours=JWT_EXPIRATION_HOURS)
    payload["exp"] = expire
    payload["iat"] = datetime.now(timezone.utc)
    return jwt.encode(payload, JWT_SECRET, algorithm=JWT_ALGORITHM)


def decode_access_token(token: str) -> dict | None:
    try:
        payload = jwt.decode(token, JWT_SECRET, algorithms=[JWT_ALGORITHM])
        return payload
    except jwt.PyJWTError:
        return None

File: backend/src/utils/auth_utils.js
import crypto from "crypto";
import bcrypt from "bcrypt";
import jwt from "jsonwebtoken";

const JWT_SECRET = process.env.JWT_SECRET || "change-me-in-production";
const JWT_ALGORITHM = "HS256";
export const JWT_EXPIRATION_HOURS = 24;

export function hashPassword(password) {
  const pwd = typeof password === "string" ? password : String(password);
  return bcrypt.hashSync(pwd, bcrypt.genSaltSync());
}

export function verifyPassword(plainPassword, hashedPassword) {
  const pwd = typeof plainPassword === "string" ? plainPassword : String(plainPassword);
  return bcrypt.compareSync(pwd, hashedPassword);
}

export function createAccessToken(data) {
  const jti = crypto.randomUUID();
  const expire = new Date();
  expire.setHours(expire.getHours() + JWT_EXPIRATION_HOURS);
  const payload = {
    ...data,
    jti,
    exp: Math.floor(expire.getTime() / 1000),
    iat: Math.floor(Date.now() / 1000),
  };
  const token = jwt.sign(payload, JWT_SECRET, { algorithm: JWT_ALGORITHM });
  return [token, jti];
}

export function decodeAccessToken(token) {
  try {
    return jwt.verify(token, JWT_SECRET, { algorithms: [JWT_ALGORITHM] });
  } catch {
    return null;
  }
}
